genealogos/src/bom/cyclonedx.rs

```diff
@@ -0,0 +1,334 @@
+            spec_version,
+            file_format,
+        })
+    }
+
+    pub fn parse(spec_version: &str, file_format: &str) -> Result<Self> {
+        let spec_version = SpecVersion::from_str(spec_version)?;
+        let file_format = FileFormat::from_str(file_format)?;
+        Ok(CycloneDX {
+            spec_version,
+            file_format,
+        })
+    }
+}
+
+impl Default for CycloneDX {
+    fn default() -> Self {
+        // TODO: Update to 1_5, or ideally Default (but that's not implemented)
+        Self::new(SpecVersion::V1_4, FileFormat::JSON)
+    }
+}
+
+impl super::Bom for CycloneDX {
+    fn write_to_io_writer<W: std::io::Write>(
+        &self,
+        model: crate::model::Model,
+        writer: &mut W,
+    ) -> crate::Result<()> {
+        // Convert the model into a CycloneDX BOM
+        let bom = Bom::try_from(model)?;
+
+        match self.spec_version {
+            SpecVersion::V1_3 => match self.file_format {
+                FileFormat::JSON => bom.output_as_json_v1_3(writer)?,
+                FileFormat::XML => bom.output_as_xml_v1_3(writer)?,
+            },
+            SpecVersion::V1_4 => match self.file_format {
+                FileFormat::JSON => bom.output_as_json_v1_4(writer)?,
+                FileFormat::XML => bom.output_as_xml_v1_4(writer)?,
+            },
+        }
+
+        Ok(())
+    }
+}
+
+impl TryFrom<Model> for Bom {
+    type Error = crate::Error;
+
+    fn try_from(model: Model) -> Result<Self> {
+        let mut model = model;
+        if cfg!(test) || std::env::var("GENEALOGOS_DETERMINISTIC").is_ok() {
+            model.components.sort_by_key(|c| c.r#ref.clone());
+
+            // This only sort the dependencies by key, not the depends on
+            // hashset, that is done later, when it is converted into a vec.
+            model.dependencies.sort_by_key(|c| c.r#ref.clone());
+        }
+
+        let mut bom = Bom {
+            components: Some(Components(
+                model
+                    .components
+                    .into_iter()
+                    .map(TryInto::try_into)
+                    .collect::<Result<Vec<_>>>()?,
+            )),
+            dependencies: Some(Dependencies(
+                model
+                    .dependencies
+                    .into_iter()
+                    .map(Into::into)
+                    .collect::<Vec<_>>(),
+            )),
+            // Constructs a BOM with a default version of 1 and serial_number with a random UUID
+            ..Default::default()
+        };
+
+        if cfg!(test) || std::env::var("GENEALOGOS_DETERMINISTIC").is_ok() {
+            bom.serial_number = Some(
+                UrnUuid::new("urn:uuid:00000000-0000-0000-0000-000000000000".to_owned()).unwrap(),
+            );
+        }
+
+        Ok(bom)
+    }
+}
+
+impl TryFrom<ModelComponent> for Component {
+    type Error = crate::Error;
+
+    fn try_from(model: ModelComponent) -> Result<Self> {
+        Ok(Component {
```

Comment:
I don't know if it's feasible, but if `Component` can derive `Default` for free, you might elide a lot of those `foo: None` parameters for free, with a `..Default::default()`. On the other hand, you might have done this on purpose to make sure everything is listed and initialized to what it's supposed to, in which case you can ignore this comment.